src/db/file.rs

```diff
@@ -5,62 +5,108 @@
 
 
-use std::path::Path;
+use std::path::{PathBuf, Path};
 use postgres::Connection;
 use rustc_serialize::json::{Json, ToJson};
-use std::fs::File;
+use std::fs;
 use std::io::Read;
 use error::Result;
 use failure::err_msg;
-
-
-fn file_path(prefix: &str, name: &str) -> String {
-    match prefix.is_empty() {
-        true => name.to_owned(),
-        false => format!("{}/{}", prefix, name),
-    }
-}
+use rusoto_s3::{S3, PutObjectRequest, GetObjectRequest, S3Client};
+use rusoto_core::region::Region;
+use rusoto_credential::EnvironmentProvider;
 
 
 fn get_file_list_from_dir<P: AsRef<Path>>(path: P,
-                                          prefix: &str,
-                                          files: &mut Vec<String>)
+                                          files: &mut Vec<PathBuf>)
                                           -> Result<()> {
     let path = path.as_ref();
 
     for file in try!(path.read_dir()) {
         let file = try!(file);
 
         if try!(file.file_type()).is_file() {
-            file.file_name().to_str().map(|name| files.push(file_path(prefix, name)));
+            files.push(file.path());
         } else if try!(file.file_type()).is_dir() {
-            file.file_name()
-                .to_str()
-                .map(|name| get_file_list_from_dir(file.path(), &file_path(prefix, name), files));
+            try!(get_file_list_from_dir(file.path(), files));
         }
     }
 
     Ok(())
 }
 
 
-pub fn get_file_list<P: AsRef<Path>>(path: P) -> Result<Vec<String>> {
+pub fn get_file_list<P: AsRef<Path>>(path: P) -> Result<Vec<PathBuf>> {
     let path = path.as_ref();
-    let mut files: Vec<String> = Vec::new();
+    let mut files = Vec::new();
 
     if !path.exists() {
         return Err(err_msg("File not found"));
     } else if path.is_file() {
-        path.file_name()
-            .and_then(|name| name.to_str())
-            .map(|name| files.push(format!("{}", name)));
+        files.push(PathBuf::from(path.file_name().unwrap()));
     } else if path.is_dir() {
-        try!(get_file_list_from_dir(path, "", &mut files));
+        try!(get_file_list_from_dir(path, &mut files));
+        for file_path in &mut files {
+            // We want the paths in this list to not be {path}/bar.txt but just bar.txt
+            *file_path = PathBuf::from(file_path.strip_prefix(path).unwrap());
+        }
     }
 
     Ok(files)
 }
 
+pub struct Blob {
+    pub path: String,
+    pub mime: String,
+    pub date_updated: time::Timespec,
+    pub content: Vec<u8>,
+}
+
+pub fn get_path(conn: &Connection, path: &str) -> Option<Blob> {
+    let rows = conn.query("SELECT path, mime, date_updated, content
+                           FROM files
+                           WHERE path = $1", &[&path]).unwrap();
+
+    if rows.len() == 0 {
+        None
+    } else {
+        let client = s3_client();
+        let row = rows.get(0);
+        let content = client.and_then(|c| c.get_object(GetObjectRequest {
```

Comment:
I wonder if we could use the `in-s3` hard-coded value to gate this check, so that we don't generate erroneous calls to S3.

Reply:
That's a great idea, it'll definitely help with the common case for now. I've pushed up a commit that does this.